lib/queue.js

```diff
@@ -60,16 +60,29 @@ function KafkaProducer(params) {
     kOpts.zkOpts = {}; // Zookeeper connection options (see: http://git.io/vCKRh)
 
     var kafka = P.promisifyAll(require('kafka-node'));
-    var client = new kafka.Client(kOpts.connectionString, kOpts.clientId, kOpts.zkOpts);
-    this.producer = new kafka.HighLevelProducer(client);
+    this.client = new kafka.Client(kOpts.connectionString, kOpts.clientId, kOpts.zkOpts);
+    this.producer = new kafka.HighLevelProducer(this.client);
 
     var self = this;
 
     // A promise that resolves only after the 'ready' event fires.
     this.ready = new P(function(resolve, reject) {
-        self.producer.on('ready', function(){
-            resolve();
-            self.logger.log('info/kafka/init', 'kafka producer is ready');
+        var isConnected = false;
+        var isReady = false;
+        function canProceed() {
+            if (isConnected && isReady) {
+                self.logger.log('info/kafka/init', 'kafka producer is ready');
+                isConnected = isReady = false;
+                resolve();
+            }
+        }
+        self.client.on('connect', function() {
+            isConnected = true;
+            canProceed();
+        });
+        self.producer.on('ready', function() {
+            isReady = true;
+            canProceed();
```

Comment:
It looks like it could be rewritten in a more compact way, something like:
```javascript
this.ready = P.join(
   new P(function(resolve, reject) { self.client.on('connect', resolve)}),
   new P(function(resolve, reject) { self.client.on('ready', resolve)})
)
.then(function() {
   self.logger.log('info/kafka/init', 'kafka producer is ready');
});
```
However I didn't test this, and if you like explicit way more, disregard this.